### ISSUE TITLE
insertOoxml() method not working in Word web version

### ISSUE THREAD

**Issue by src107** [OfficeDev/office-js#3271]
insertOoxml() method not working in Word web version, when trying to run it from ScriptLab with the examples given in the documentation of office-js.
PFB the error : -
`{
    "stack": "RichApi.Error: unknown\n    at new n ([https://appsforoffice.microsoft.com/lib/1/hosted/word-web-16.00.js:25:330439)\n](https://appsforoffice.microsoft.com/lib/1/hosted/word-web-16.00.js:25:330439)/n)    at o.processRequestExecutorResponseMessage ([https://appsforoffice.microsoft.com/lib/1/hosted/word-web-16.00.js:25:394605)\n](https://appsforoffice.microsoft.com/lib/1/hosted/word-web-16.00.js:25:394605)/n)    at [https://appsforoffice.microsoft.com/lib/1/hosted/word-web-16.00.js:25:392668"](https://appsforoffice.microsoft.com/lib/1/hosted/word-web-16.00.js:25:392668%22),
    "message": "unknown",
    "name": "RichApi.Error",
    "code": "GeneralException",
    "traceMessages": [],
    "innerError": null,
    "debugInfo": {
        "code": "GeneralException",
        "message": "unknown",
        "errorLocation": "Range.insertOoxml",
        "statement": "var insertOoxml = selection.insertOoxml(...);",
        "surroundingStatements": [
            "var root = context.root;",
            "var selection = root.getSelection();",
            "// Instantiate {selection}",
            "// >>>>>",
            "var insertOoxml = selection.insertOoxml(...);",
            "// <<<<<",
            "// Instantiate {insertOoxml}"
        ],
        "fullStatements": [
            "Please enable config.extendedErrorLogging to see full statements."
        ]
    },
    "httpStatusCode": 500
}`

## Your Environment
* Platform :  Office on the web
* Host : Word
* Office version/build number: 16.0.16330.41013
* Session ID : 6e297e34-ddca-4ba2-b740-e79046b5f4dc
* Operating System: Mac, Windows
* Browser (if using Office on the web): Chrome, Safari, Firefox
 

## Steps to reproduce

1. Open Script Lab
2. Execute the example code given the documentation ([link](https://learn.microsoft.com/en-us/javascript/api/word/word.body?view=word-js-preview#word-word-body-insertooxml-member(1))

## Link to live example(s)
PFB the code snippet :-
`async function run() {
  await Word.run(async (context) => {
    const body = context.document.body;

    // Queue a command to insert OOXML in to the beginning of the body.
    body.insertOoxml("<pkg:package xmlns:pkg='http://schemas.microsoft.com/office/2006/xmlPackage'><pkg:part pkg:name='/_rels/.rels' pkg:contentType='application/vnd.openxmlformats-package.relationships+xml' pkg:padding='512'><pkg:xmlData><Relationships xmlns='http://schemas.openxmlformats.org/package/2006/relationships'><Relationship Id='rId1' Type='http://schemas.openxmlformats.org/officeDocument/2006/relationships/officeDocument' Target='word/document.xml'/></Relationships></pkg:xmlData></pkg:part><pkg:part pkg:name='/word/document.xml' pkg:contentType='application/vnd.openxmlformats-officedocument.wordprocessingml.document.main+xml'><pkg:xmlData><w:document xmlns:w='http://schemas.openxmlformats.org/wordprocessingml/2006/main' ><w:body><w:p><w:pPr><w:spacing w:before='360' w:after='0' w:line='480' w:lineRule='auto'/><w:rPr><w:color w:val='70AD47' w:themeColor='accent6'/><w:sz w:val='28'/></w:rPr></w:pPr><w:r><w:rPr><w:color w:val='70AD47' w:themeColor='accent6'/><w:sz w:val='28'/></w:rPr><w:t>This text has formatting directly applied to achieve its font size, color, line spacing, and paragraph spacing.</w:t></w:r></w:p></w:body></w:document></pkg:xmlData></pkg:part></pkg:package>", Word.InsertLocation.start);

    // Synchronize the document state by executing the queued commands,
    // and return a promise to indicate task completion.
    await context.sync();
    console.log('OOXML added to the beginning of the document body.');
  });
}
`

Script lab code : https://gist.github.com/src107/92c6ea11eb98062c8b91d5b858d941ac

**Comment by penglongzhaochina** [comment]
Thanks for reporting this issue.

It has been put on our backlog<Bug#7765252> for internal track. We will keep track of this issue and let you know if there are any updates.

**Comment by CampLegal-Govind** [comment]
I face similar error while working word add in.
`word-web-16.00.js:25 Uncaught (in promise) RichApi.Error: ooxmlIsMalformated
    at new n (word-web-16.00.js:25:330439)
    at o.processRequestExecutorResponseMessage (word-web-16.00.js:25:394605)
    at word-web-16.00.js:25:392668`

**Comment by src107** [comment]
Hi @penglongzhaochina ,
There is a build number change for the web version which is "**16.0.16403.41005**". 
insertOoxml() method is working but the changes in the document are reflecting after some time (average time of around 10 secs).

Please try the sample code from the documentation : https://gist.github.com/src107/92c6ea11eb98062c8b91d5b858d941ac
Session Id: 24326398-95ab-48b6-a4e1-cdcc091c4550

Let us know what could be the issue or is the team looking into it?

**Comment by JinghuiMS** [comment]
Thanks for reporting this issue in time. Sorry for the inconvenience caused to you. These two days we are actively working on this. The fix should be available to all of you. Please double check it in your environment after clearing your browser cache.
Please let us know if you have any further questions. Thank you all again!

**Comment by src107** [comment]
Hi @JinghuiMS ,
Thank you for the reply, insertOoxml is now working in the web version but having some performance implications i.e Its reflecting in the document after around 10 to 15 secs. Is anyone looking into this issue too?

I have tried inserting the documentation code: https://gist.github.com/src107/92c6ea11eb98062c8b91d5b858d941ac

**Comment by JinghuiMS** [comment]
@src107 For the new issue you mention this time, we have already created an internal work item<7776847> on it. We will keep track of this issue and let you know if there are any updates. Thank you!

**Comment by Tebogo11** [comment]
Hi is theirs any updates, 
experiencing a similar issue with insertOoxml on Word online, when running the example up top the ooxml is inserted with a 10 sec delay, but when inserting a much larger ooxml it's reverting back to the error message with no insert

**Comment by JinghuiMS** [comment]
The team will review all the work items in our backlog regularly. They will priority the work item. We unfortunately have no timelines to share at this point.

**Comment by marcinkapusta** [comment]
@JinghuiMS is there any update on this?
Is it possible to increase the priority of this issue?

In our scenario, it takes about 30-40 seconds for the inserted ooxml to appear in the document.
Users feel as if it doesn't work

Everything works very well in the desktop edition, therefore it prevents our users from migrating to Word online (at least those who need our add-ins).

**Comment by xiaoyuMS** [comment]
Hi @marcinkapusta, thanks for reaching us. 
Is it possible to know how many users have been affected? We are currently working on a fix for this issue and will return back to this thread once the fix is ready.

**Comment by marcinkapusta** [comment]
Hi @xiaoyuMS , thanks for your response.

It is difficult to estimate the exact amount.
Word add-in is available to all our clients (together with the platform we offer). We do not collect statistics on how many of them actively use the add-in, how many of them use the Desktop version, and how many of them use the online version.

Currently, we have received information from one of our clients that some of the functionality does not work correctly in the online version and we have reproduced the problem in our environments.

It appears that any Word online add-in user will be affected by the problem. It would be great to have the problem fixed before other clients start reporting it.

**Comment by xiaoyuMS** [comment]
Hi @marcinkapusta @Tebogo11 @src107  @CampLegal-Govind  , the fix has been deployed to production, please have a try. The full deployment may take another 1-2 days.

**Comment by xiruatms** [comment]
@marcinkapusta @Tebogo11 @src107 @CampLegal-Govind , will you please take a try with it? The fix is ready on production now.

**Comment by marcinkapusta** [comment]
Hi @xiaoyuMS and @xiruatms .
I'm sorry for the late answer.

We verified the fix in the mid-November already.
In our scenario, it takes max 4 seconds for the inserted ooxml to appear in the document (it was 30-40 seconds before).
During this time there is also an additional message informing the user that something is still being processed. 

The solution is not perfect as in our scenario there are two different messages one after another, but at least the user is not confused that nothing is happening.
 
Thank you for the fix.

**Comment by src107** [comment]
> @marcinkapusta @Tebogo11 @src107 @CampLegal-Govind , will you please take a try with it? The fix is ready on production now.

Hi @xiruatms and @xiaoyuMS ,
I had verified earlier the performance has improved, but today while testing I found that ooxml insertion is not happening, it is throwing some GeneralApi exception.

<img width="656" alt="image" src="https://github.com/OfficeDev/office-js/assets/49275977/32f23cd8-4084-4a9c-ab7a-f271cca63a84">

Build: 16.0.17213.41019
Session Id: eda1b5e5-7469-48e7-92ca-fcc80165e41f

**Comment by xiruatms** [comment]
@marcinkapusta , thanks for your confirmation. We'll further investigate how to improve the performance. It might take a while. Thank you for your patience.

**Comment by xiruatms** [comment]
@src107 , thank you for taking a try. May I have more information about the failure you encountered? Can you share with us the testing steps and code snippet with us?

**Comment by src107** [comment]
> @src107 , thank you for taking a try. May I have more information about the failure you encountered? Can you share with us the testing steps and code snippet with us?

Hi @xiruatms,
I am trying the code given in the documentation only from ScriptLab
https://learn.microsoft.com/en-us/javascript/api/word/word.contentcontrol?view=word-js-preview#word-word-contentcontrol-insertooxml-member(1)

**Comment by xiruatms** [comment]
@src107 , thanks for sharing.

I took a try and it worked as expected as following:
![image](https://github.com/OfficeDev/office-js/assets/40648203/d075e8f3-f049-422b-9b07-9688b3ee891d)

Will you please let me know: how did you create your contentcontrol in this case? Can you still reproduce the general exception you mentioned consistently?

**Comment by src107** [comment]
> @src107 , thanks for sharing.
> 
> I took a try and it worked as expected as following: ![image](https://private-user-images.githubusercontent.com/40648203/292652225-d075e8f3-f049-422b-9b07-9688b3ee891d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDM0MTQxMjMsIm5iZiI6MTcwMzQxMzgyMywicGF0aCI6Ii80MDY0ODIwMy8yOTI2NTIyMjUtZDA3NWU4ZjMtZjA0OS00MjJiLTliMDctOTY4OGIzZWU4OTFkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMjQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjI0VDEwMzAyM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPThhOWIxYWJiMThjZjk5NmNjNjk4ZjYyOGVhNjhjZmMyYzAwYTU2ZDJkNDY2ZWFhMmM2MTVhNTEwNzFjZTczNWQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.hz5uH9jfKDkysxHu0Fu5pdT6jpX9rTyADBsmRxWxb5o)
> 
> Will you please let me know: how did you create your contentcontrol in this case? Can you still reproduce the general exception you mentioned consistently?

Hi @xiruatms,

I used the below code and the issue is still reproducible, this is happening in the web version, but works fine in desktop version.

`$("#run").click(() => tryCatch(run));
async function run() {
  await Word.run(async (context) => {
    context.document.getSelection().insertContentControl();
    await context.sync();

    const contentControls = context.document.contentControls;

    // Queue a command to load the id property for all of the content controls.
    contentControls.load('id');

    // Synchronize the document state by executing the queued commands, 
    // and return a promise to indicate task completion.
    await context.sync();
    if (contentControls.items.length === 0) {
      console.log('No content control found.');
    }
    else {
      // Queue a command to put OOXML into the contents of the first content control.
      contentControls.items[0].insertOoxml("<pkg:package xmlns:pkg='http://schemas.microsoft.com/office/2006/xmlPackage'><pkg:part pkg:name='/_rels/.rels' pkg:contentType='application/vnd.openxmlformats-package.relationships+xml' pkg:padding='512'><pkg:xmlData><Relationships xmlns='http://schemas.openxmlformats.org/package/2006/relationships'><Relationship Id='rId1' Type='http://schemas.openxmlformats.org/officeDocument/2006/relationships/officeDocument' Target='word/document.xml'/></Relationships></pkg:xmlData></pkg:part><pkg:part pkg:name='/word/document.xml' pkg:contentType='application/vnd.openxmlformats-officedocument.wordprocessingml.document.main+xml'><pkg:xmlData><w:document xmlns:w='http://schemas.openxmlformats.org/wordprocessingml/2006/main' ><w:body><w:p><w:pPr><w:spacing w:before='360' w:after='0' w:line='480' w:lineRule='auto'/><w:rPr><w:color w:val='70AD47' w:themeColor='accent6'/><w:sz w:val='28'/></w:rPr></w:pPr><w:r><w:rPr><w:color w:val='70AD47' w:themeColor='accent6'/><w:sz w:val='28'/></w:rPr><w:t>This text has formatting directly applied to achieve its font size, color, line spacing, and paragraph spacing.</w:t></w:r></w:p></w:body></w:document></pkg:xmlData></pkg:part></pkg:package>", "End");

      // Synchronize the document state by executing the queued commands, 
      // and return a promise to indicate task completion.
      await context.sync();
      console.log('Inserted OOXML in the first content control.');
    }
  }); 
}

/** Default helper for invoking an action and handling errors. */
async function tryCatch(callback) {
  try {
    await callback();
  } catch (error) {
    // Note: In a production add-in, you'd want to notify the user through your add-in's UI.
    console.error(error);
  }
}
`

**Comment by xiruatms** [comment]
@src107 , thanks for all the information. Unfortunately, I can't reproduce this with the codes you provided:
![image](https://github.com/OfficeDev/office-js/assets/40648203/f5e1944e-c482-4fd2-a860-29295178a760)

and I'm testing with the same build as yours:
![image](https://github.com/OfficeDev/office-js/assets/40648203/875bd4cc-9fb4-4b2a-bce5-7f69be141947)

If possible, can you please take a screen record for us in case we missed anything in your verification? Does this happen with an existing document or a new document? Can you share with us your document if this happened with a specific document? That'll be much helpful for us to continue investigation.

**Comment by src107** [comment]
Hi @xiruatms , the insertOoxml method is working fine, but there is one issue. When I try to insert ooxml objects inside a content control, it works fine the first time. However, the second time, the content control of the first ooxml object gets deleted.

This is only happening in web version
Build: 16.0.17308.41004

Code [Link](https://gist.github.com/src107/63759e3d6d35eea3f9820b9efcefbac0)
Please try to run the above code to reproduce the issue.

**Comment by src107** [comment]
> Hi @xiruatms , the insertOoxml method is working fine, but there is one issue. When I try to insert ooxml objects inside a content control, it works fine the first time. However, the second time, the content control of the first ooxml object gets deleted.
> 
> This is only happening in web version Build: 16.0.17308.41004
> 
> Code [Link](https://gist.github.com/src107/63759e3d6d35eea3f9820b9efcefbac0) Please try to run the above code to reproduce the issue.

Hi @xiruatms, anything on the above issue?

**Comment by xiruatms** [comment]
@src107 , thanks for your patience. 

Just want to confirm the meaning of "the second time" in your question. Does that mean that: clear the content control which was inserted "the first time"; and then run your code again for "the second time". Correct?

**Comment by aeturneus** [comment]
We are seeing errors for `insertOoxml` in our WOPI environment
@xiruatms we're also seeing insertOoxml fail in WOPI, again, after it was working for a while. The error message is:

Uncaught (in promise) RichApi.Error: The action isn't supported by Word in a browser. Check the OfficeExtension.Error.debugInfo for more information. at new n (word-web-16.00.js:25:340817) at o.processRequestExecutorResponseMessage (word-web-16.00.js:25:405264) at word-web-16.00.js:25:403327 at async t.default.unlockContentControls (taskpane.js:393:427752)

The full console statement is:

{ "code": "NotAllowed", "message": "The action isn't supported by Word in a browser. Check the OfficeExtension.Error.debugInfo for more information.", "errorLocation": "Body.insertOoxml", "statement": "var insertOoxml = body.insertOoxml(...);", "surroundingStatements": [ "var root = context.root;", "var body = root.body;", "// Instantiate {body}", "// >>>>>", "var insertOoxml = body.insertOoxml(...);", "// <<<<<", "// Instantiate {insertOoxml}" ], "fullStatements": [ "Please enable config.extendedErrorLogging to see full statements." ] }

The same code works fine in desktop, but fails in WOPI when running with our add-in. Cross-posting from https://github.com/OfficeDev/office-js/issues/2840 as this seems to be a related issue.

**Comment by xiruatms** [comment]
@aeturneus , feel free to create a new GitHub issue to address your errors. And please provide details (testing environment, testing document, codes snippet) as much as possible in that issue so that we can follow up. Thanks.
